Ladybird/Android/src/main/java/org/serenityos/ladybird/LadybirdActivity.kt

```diff
@@ -28,16 +31,24 @@ class LadybirdActivity : AppCompatActivity() {
         binding = ActivityMainBinding.inflate(layoutInflater)
         setContentView(binding.root)
         setSupportActionBar(binding.toolbar)
+        urlEditText = binding.urlEditText
         view = binding.webView
+        view.onLoadStart = { url: String, _ ->
+            urlEditText.setText(url, TextView.BufferType.EDITABLE)
+        }
+        urlEditText.setOnEditorActionListener { textView: TextView?, i: Int, _: KeyEvent? ->
+            Boolean
+            if (i == EditorInfo.IME_ACTION_GO || i == EditorInfo.IME_ACTION_SEARCH) {
+                view.loadURL(textView!!.text.toString())
+            }
+            false
+        }
```

Comment:
@bugaevc Is doing all of this manual wiring in your activity's onCreate the idiomatic way to do this? I suppose for the moment it's not a big deal but i can see this getting crowded. I assume the public Function callback thing we do in LibGUI/LibWeb isn't exactly the idiomatic way either :sweat_smile:

Reply:
I think so; `onCreate()` typically grows to be the largest method.

In general, there's this whole theory about how (big) Android apps should be structured, with MVVM, repository and UI layers, and whatnot. https://developer.android.com/topic/architecture/recommendations Ladybird is too small for that for now, but we'll get there once you'd want to store history and bookmarks in a persistent store (if you're planning to use Android's [SQLite with Room](https://developer.android.com/training/data-storage/room), and not LibSQL — that is).

Reply:
All right. I'll probably split it into some helpers eventually. 

I'm not quite sure whether I'm following those split of View, ViewModel, Data, etc. I don't think the app does? Things like "never call into the UI from the ViewModel, just update the state" is... not what I'm doing at the moment. But it doesn't seem like it'll be too difficult (?) to make that a thing.

As far as Bookmarks and history, We've talked about this a lot already I think. It would be my preference to just create a wrapper around our C++ objects and use those as the data model for those items. If that becomes a giant mess then we can look at doing a chrome-specific impl. but that's for sure not preferred.

Reply:
I think we'd be able to have `LiveData`-style wrappers around LibSQL data and slot that in where SQLite/Room would go. We'd need some way for LibSQL to notify users that data has changed though, which I assume we don't have.